src/core/Sample.cpp

```diff
@@ -117,58 +114,40 @@ auto Sample::operator=(Sample&& other) -> Sample&
 	return *this;
 }
 
-bool Sample::play(sampleFrame* dst, PlaybackState* state, int numFrames, float desiredFrequency, Loop loopMode)
+bool Sample::play(sampleFrame* dst, PlaybackState* state, size_t numFrames, float desiredFrequency, Loop loopMode) const
 {
-	if (numFrames <= 0 || desiredFrequency <= 0) { return false; }
+	if (loopMode == Loop::Off && (state->m_frameIndex >= m_endFrame || (state->m_frameIndex < 0 && state->m_backwards)))
+	{
+		return false;
+	}
+
+	const auto outputSampleRate = Engine::audioEngine()->processingSampleRate() * m_frequency / desiredFrequency;
+	const auto inputSampleRate = m_buffer->sampleRate();
+	const auto resampleRatio = outputSampleRate / inputSampleRate;
+	const auto marginSize = s_interpolationMargins[state->resampler().interpolationMode()];
 
-	auto resampleRatio = static_cast<float>(Engine::audioEngine()->processingSampleRate()) / m_buffer->sampleRate();
-	resampleRatio *= frequency() / desiredFrequency;
+	auto playBuffer = std::vector<sampleFrame>(numFrames / resampleRatio + marginSize);
+	playRaw(playBuffer.data(), playBuffer.size(), state, loopMode);
 
-	auto playBuffer = std::vector<sampleFrame>(numFrames / resampleRatio);
-	if (!typeInfo<float>::isEqual(resampleRatio, 1.0f))
+	const auto resampleResult
+		= state->resampler().resample(&playBuffer[0][0], playBuffer.size(), &dst[0][0], numFrames, resampleRatio);
+	advance(state, resampleResult.inputFramesUsed, loopMode);
+
+	if (resampleResult.outputFramesGenerated < numFrames)
 	{
-		playBuffer.resize(playBuffer.size() + s_interpolationMargins[state->resampler().interpolationMode()]);
+		std::fill(dst + resampleResult.outputFramesGenerated, dst + resampleResult.outputFramesGenerated + numFrames,
+			sampleFrame{0, 0});
```

Comment:
I feel like this is more readable.
```suggestion
		std::fill(dst + resampleResult.outputFramesGenerated,
			dst + resampleResult.outputFramesGenerated + numFrames,
			sampleFrame{0, 0});
```

Reply:
Or better yet, use `std::fill_n`